doc/manual/metaprogramming.rst

```diff
@@ -872,3 +872,194 @@ entirely in Julia. You can read their source and see precisely what they
+
+Since the body of the staged function is non-deterministic, its behavior
+is undefined; the expression returned on the *first* invocation will be
+used for *all* subsequent invocations with the same type. When we call
+the staged function with ``x`` of a new type, ``rand()`` will be called
+again to see which method body to use for the new type. In this case, for
+one *type* out of ten, ``baz(x)`` will return the string ``"boo!"``.
+In short: don't do this.
+
+While these examples are perhaps not so interesting, they have hopefully
+helped to illustrate how staged functions work, both in the definition end
+and at the call site. Next, let's build some more advanced functionality
+using staged functions...
+
+An advanced example
+~~~~~~~~~~~~~~~~~~~
+
+Julia's base library has a function ``sub2ind`` function to calculate a
+linear index into an n-dimensional array, based on a set of n multilinear
+indices - in other words, to calculate the index ``i`` that can be used to
+index into an array ``A`` using ``A[i]``, instead of ``A[x,y,z,...]``. One
+possible implementation is the following::
+
+    function sub2ind_loop(dims::NTuple{N}, I::Integer...)
+        ind = I[N] - 1
+        for i = N-1:-1:1
+            ind = I[i]-1 + dims[i]*ind
+        end
+        ind + 1
+    end
+
+The same thing can be done using recursion::
+
+    sub2ind_rec(dims::()) = 1
+    sub2ind_rec(dims::(),i1::Integer, I::Integer...) =
+        i1==1 ? sub2ind(dims,I...) : throw(BoundsError())
+    sub2ind_rec(dims::(Integer,Integer...), i1::Integer) = i1
+    sub2ind_rec(dims::(Integer,Integer...), i1::Integer, I::Integer...) =
+        i1 + dims[1]*(sub2ind(tail(dims),I...)-1)
+
+Both these implementations, although different, do essentially the same
+thing: a runtime loop over the dimensions of the array, collecting the
+offset in each dimension into the final index.
+
+However, all the information we need for the loop is embedded in the type
+information of the arguments. Thus, we can utilize staged functions to
+move the iteration to compile-time. The body becomes almost identical,
+but instead of calculating the linear index, we build up an *expression*
+that calculates the index::
+
+    stagedfunction sub2ind_staged{N}(dims::NTuple{N}, I::Integer...)
+        ex = :(I[$N] - 1)
+        for i = N-1:-1:1
+            ex = :(I[$i] - 1 + dims[$i]*$ex)
+        end
+        :($ex + 1)
+    end
+
+**What code will this staged function generate?**
+
+An easy way to find out, is to extract the body into another (regular)
+function::
+
+    stagedfunction sub2ind_staged{N}(dims::NTuple{N}, I::Integer...)
+        sub2ind_staged_impl(dims, I...)
+    end
+
+    function sub2ind_staged_impl(dims, I...)
+        ex = :(I[$N] - 1)
+        for i = N-1:-1:1
+            ex = :(I[$i] - 1 + dims[$i]*$ex)
+        end
+        :($ex + 1)
+    end
+
+We can now execute ``sub2ind_staged_impl`` and examine the expression it
+returns::
+
+    julia> sub2ind_staged_impl((5,2), 3, 2)
```

Comment:
This should probably be called with the types — like the staged function would do: `sub2ind_staged_impl((Int,Int), Int, Int).

Also, `N` is not defined in the `_impl` method.